### PR TITLE
v1.14 Backports 2023-11-29

### PR DESCRIPTION
**Pull request by joamaki**
* [x] #29178 (@brb)
 * [x] #29189 (@brb)
 * [x] #29020 (@jrajahalme)
 * [x] #29273 (@nebril)
 * [x] #29160 (@julianwiedmann) ⚠️ adapted code, review carefully
 * [x] #29325 (@brb)
 * [x] #29348 (@julianwiedmann)
 * [x] #28813 (@oblazek) ⚠️ few conflicts, review carefully
 * [x] #29367 (@mhofstetter)
 * [x] #29352 (@pchaigno)
 * [x] #29409 (@giorio94)
 * [x] #29353 (@pchaigno)
 * [x] #29310 (@julianwiedmann)
 * [x] #29378 (@giorio94)

PRs skipped due to conflicts:

 * #29162 (@jrajahalme)
 *  #29155 (@chez-shanpu) 

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 27007 29178 29189 29020 29273 29160 29325 29348 28813 29367 29352 29409 29353 29310 29378
```

### CODE REVIEW

**Review by mhofstetter — APPROVED** [review]
Thanks @joamaki

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -1134,12 +1137,10 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
...
 #if (defined(IS_BPF_OVERLAY) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE) || \
 	(!defined(IS_BPF_OVERLAY) && DSR_ENCAP_MODE != DSR_ENCAP_GENEVE)
-			bool dsr = false;
-
 			ret = nodeport_extract_dsr_v6(ctx, ip6, &tuple, l4_off,
 						      &key.address,
-						      &key.dport, &dsr);
-			if (dsr) {
+						      &key.dport, dsr);
+			if (*dsr) {
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]: Just spotted this while driving-by, and think the approach we took in `main` won't work for `v1.14`. Probably best to drop https://github.com/cilium/cilium/pull/29155 for now, sorry :/.

`main` has https://github.com/cilium/cilium/pull/27267 and returns inline. `v1.14` returns via tail-call, so the `dsr` flag will never propagate back.

@ysksuzuki @chez-shanpu fyi. Think we need to check if we can build a custom backport.

**Review by oblazek — APPROVED** [review]
Great, thanks.

**Review by giorio94 — APPROVED** [review]
My commits look good. Thanks!

**Comment by julianwiedmann** [comment]
>  [bpf: nat: only set SNAT_DONE when packet was actually SNATed #29160](https://github.com/cilium/cilium/pull/29160) (@julianwiedmann) ⚠️ adapted code, review carefully

:pray: let's try the following. Ideally that should also apply to v1.12 and v1.13 ... will check.

```
diff --git a/bpf/lib/nodeport.h b/bpf/lib/nodeport.h
index 53d2639238..11e08580f1 100644
--- a/bpf/lib/nodeport.h
+++ b/bpf/lib/nodeport.h
@@ -165,14 +165,17 @@ static __always_inline int nodeport_snat_fwd_ipv6(struct __ctx_buff *ctx,
        bool snat_needed;
 
        snat_needed = snat_v6_prepare_state(ctx, &target);
-       if (snat_needed)
+       if (snat_needed) {
                ret = snat_v6_nat(ctx, &target, trace, ext_err);
+
+               /* See the equivalent v4 path for comment */
+               if (!IS_ERR(ret))
+                       ctx_snat_done_set(ctx);
+       }
+
        if (ret == NAT_PUNT_TO_STACK)
                ret = CTX_ACT_OK;
 
-       /* See the equivalent v4 path for comment */
-       ctx_snat_done_set(ctx);
-
        return ret;
 }
 
@@ -1564,17 +1567,20 @@ static __always_inline int nodeport_snat_fwd_ipv4(struct __ctx_buff *ctx,
        bool snat_needed;
 
        snat_needed = snat_v4_prepare_state(ctx, &target);
-       if (snat_needed)
+       if (snat_needed) {
                ret = snat_v4_nat(ctx, &target, trace, ext_err);
+
+               /* If multiple netdevs process an outgoing packet, then this packets will
+                * be handled multiple times by the "to-netdev" section. This can lead
+                * to multiple SNATs. To prevent from that, set the SNAT done flag.
+                */
+               if (!IS_ERR(ret))
+                       ctx_snat_done_set(ctx);
+       }
+
        if (ret == NAT_PUNT_TO_STACK)
                ret = CTX_ACT_OK;
 
-       /* If multiple netdevs process an outgoing packet, then this packets will
-        * be handled multiple times by the "to-netdev" section. This can lead
-        * to multiple SNATs. To prevent from that, set the SNAT done flag.
-        */
-       ctx_snat_done_set(ctx);
-
        return ret;
 }
```

**Review by pchaigno — APPROVED** [review]
My two PRs look good. Thanks Jussi!

**Review by brb — APPROVED** [review]
My first two PRs seem to be no-op :thinking: . The 3rd is LGTM, thanks.

**Review by nebril — APPROVED** [review]
Thanks!

**Comment by ysksuzuki** [comment]
#27007 has been backported on https://github.com/cilium/cilium/pull/29270

File: bpf/lib/nodeport.h (hunk 1 above)
Comment (ysksuzuki) [reply bpf/lib/nodeport.h]: Ah, right. Will look into that.

Comment (ysksuzuki) [reply bpf/lib/nodeport.h]: Let's drop #29155 from this PR. I will raise a PR against v1.14 later. Is that OK with you? Or are you planning to send a PR? @chez-shanpu

Comment (chez-shanpu) [reply bpf/lib/nodeport.h]: > Let's drop #29155 from this PR. I will raise a PR against v1.14 later. Is that OK with you? Or are you planning to send a PR? @chez-shanpu

@ysksuzuki No problem, thank you :)

Comment (joamaki) [reply bpf/lib/nodeport.h]: Dropped.

**Review by jrajahalme — APPROVED** [review]

**Comment by julianwiedmann** [comment]
/test-backport-1.14

**Review by julianwiedmann — APPROVED** [review]
ty Jussi, looks good!

**Pull request merged**